### PR TITLE
[CHAIN-47] make AssetVault.sol

### PR DESCRIPTION
**Pull request by eyqs**
## What's new in this PR?

Add `AssetVault.sol` to redistribute yield from asset tokens.

Also rename `expiresAt` to `expiration`.

## Why?

What problem does this solve?
Why is this important?
What's the context?

### CODE REVIEW

File: smart-wallets/src/extensions/AssetVault.sol [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+    /**
+     * @notice Returns the locked asset token balance for the given asset token
+     * @param assetToken The asset token address
+     * @return lockedBalance The locked asset token balance
+     */
+    function getLockedAssetTokenBalance(address assetToken) public view returns (uint256 lockedBalance) {
+        YieldDistributionListItem storage distribution = _getAssetVaultStorage().yieldDistributions[assetToken];
+        uint256 lockedAmount = distribution.yield.amount;
+
+        while (lockedAmount > 0) {
```
Comment (Purva-Chaudhari) [inline smart-wallets/src/extensions/AssetVault.sol]: Are the iterations always a small number here, or it can be large too?

Comment (eyqs) [reply smart-wallets/src/extensions/AssetVault.sol]: It can be a large number too.

File: smart-wallets/src/extensions/AssetVault.sol [hunk 2]
```diff
@@ -0,0 +1,386 @@
...
+    * @param invalidUser Address of the caller who tried to call a wallet-only function
+    */
+    error UnauthorizedCall(address invalidUser);
+
+    // Modifiers
+
+    /// @notice Only the user wallet can call this function
+    modifier onlyWallet() {
+        if (msg.sender != wallet) {
+            revert UnauthorizedCall();
```
Comment (Purva-Chaudhari) [inline smart-wallets/src/extensions/AssetVault.sol]: `UnauthorizedCall()` should have `invalidUser` address

**Review by Purva-Chaudhari — APPROVED** [review]
LGTM, 

For the loops, if they get too big in iterating we can declare max gas cost for the function call if needed

**Pull request merged**